athena/kernel/src/arch/x64/context.rs

```diff
@@ -0,0 +1,317 @@
+                store as *mut ContextVariant,
+            )
+        }
+    }
+
+    /// Performs a context switch that doesn't restore.
+    ///
+    /// # Safety
+    ///
+    /// * The `restore` pointer must be a valid `Context.
+    pub unsafe fn jump(restore: *const Self) -> ! {
+        // SAFETY: repr(transparent)
+        unsafe {
+            ContextVariant::jump(restore as *const ContextVariant);
+        }
+    }
+}
+
+impl ContextVariant {
+    /// Constructs a new kernel thread context.
+    pub fn kthread<F>(f: F) -> Self
+    where
+        F: FnOnce() + Send + 'static,
+    {
+        Self::KThread(KThread::new(f))
+    }
+    /// Performs a context switch.
+    ///
+    /// The `restore` context will be restored and the current context will be
+    /// stored to `store`.
+    ///
+    /// # Safety
+    ///
+    /// * The `restore` and `store` pointers must be valid `Context`s.
+    pub unsafe fn switch(restore: *const Self, store: *mut Self) {
+        let (restore, store) = {
+            let restore = unsafe { &*restore };
+            let store = unsafe { &mut *store };
+
+            let reg_restore = match restore {
+                Self::KThread(kt) => &kt.regs,
+            };
+            let reg_store = match store {
+                Self::KThread(kt) => &mut kt.regs,
+            };
+            (reg_restore, reg_store)
+        };
+        unsafe {
+            Regs::switch(restore, store);
+        }
+    }
+
+    /// Performs a context switch that doesn't restore.
+    ///
+    /// # Safety
+    ///
+    /// * The `restore` pointer must be a valid `Context.
+    pub unsafe fn jump(restore: *const Self) -> ! {
+        let restore = {
+            let restore = unsafe { &*restore };
+
+            let reg_restore = match restore {
+                Self::KThread(kt) => &kt.regs,
+            };
+
+            reg_restore
+        };
+        unsafe {
+            Regs::jump(restore);
+        }
+    }
+}
+
+impl KThread {
+    /// Constructs the kernel thread context.
+    pub fn new<F>(f: F) -> Self
+    where
+        F: FnOnce() + Send + 'static,
+    {
+        extern "sysv64" fn inner<F>(func: *mut F) -> !
+        where
+            F: FnOnce() + Send + 'static,
+        {
+            // SAFETY: We leaked it when we created the kthread.
+            {
+                let func = unsafe { Box::from_raw(func) };
+                func();
+            }
+            sched::terminate();
+        }
+        let stack_page = mm::alloc_page().unwrap();
+        let func = Box::into_raw(Box::new(f));
+        let mut regs = Regs::new();
+        // System-V ABI pushes int-like arguements to registers.
+        regs.scratch.rdi = func as u64;
+        regs.rsp = stack_page.start_address().as_u64() + Size4KiB::SIZE;
+        regs.rip = inner::<F> as u64;
+        regs.rflags = 2;
+        Self {
+            regs,
+            _stack_page: stack_page,
+        }
+    }
+}
+
+// TODO: Drop.
```

Comment:
Maybe better solved with an AllocatedPage struct that frees the page on drop instead.